File: aws_batch_squared_py/iam_stack.py
"""IAM nested stack — roles and instance profile for AWS Batch.

Ported from ``lib/iam-stack.ts``.
"""

import aws_cdk as cdk
from aws_cdk import aws_iam as iam
from constructs import Construct


class IamStack(cdk.NestedStack):
    batch_job_role: iam.Role
    batch_instance_role: iam.Role
    batch_instance_profile: iam.InstanceProfile
    batch_spot_fleet_role: iam.Role
    batch_service_role: iam.Role

    def __init__(
        self,
        scope: Construct,
        construct_id: str,
        *,
        s3_bucket_name: str,
        **kwargs,
    ) -> None:
        super().__init__(scope, construct_id, **kwargs)

        region = cdk.Stack.of(self).region

        # Batch Job Role
        self.batch_job_role = iam.Role(
            self,
            "BatchJobRole",
            assumed_by=iam.ServicePrincipal("ecs-tasks.amazonaws.com"),
            inline_policies={
                f"S3Bucket-Access-{region}": iam.PolicyDocument(
                    statements=[
                        iam.PolicyStatement(
                            effect=iam.Effect.DENY,
                            resources=[f"arn:aws:s3:::{s3_bucket_name}"],
                            actions=["s3:Delete*", "s3:PutBucket*"],
                        ),
                        iam.PolicyStatement(
                            effect=iam.Effect.ALLOW,
                            resources=[f"arn:aws:s3:::{s3_bucket_name}"],
                            actions=["s3:ListBucket*"],
                        ),
                        iam.PolicyStatement(
                            effect=iam.Effect.ALLOW,
                            resources=[f"arn:aws:s3:::{s3_bucket_name}/*"],
                            actions=["s3:*"],
                        ),
                    ]
                ),
            },
        )

        # Batch Instance Role
        self.batch_instance_role = iam.Role(
            self,
            "BatchInstanceRole",
            assumed_by=iam.ServicePrincipal("ec2.amazonaws.com"),
            managed_policies=[
                iam.ManagedPolicy.from_aws_managed_policy_name(
                    "service-role/AmazonEC2ContainerServiceforEC2Role"
                ),
                iam.ManagedPolicy.from_aws_managed_policy_name("AmazonS3ReadOnlyAccess"),
                iam.ManagedPolicy.from_aws_managed_policy_name(
                    "AmazonSSMManagedInstanceCore"
                ),
                iam.ManagedPolicy.from_aws_managed_policy_name(
                    "CloudWatchAgentServerPolicy"
                ),
            ],
            inline_policies={
                f"S3Bucket-Access-{region}": iam.PolicyDocument(
                    statements=[
                        iam.PolicyStatement(
                            sid="S3BucketAllowAllObjectOps",
                            effect=iam.Effect.ALLOW,
                            resources=[
                                f"arn:aws:s3:::{s3_bucket_name}",
                                f"arn:aws:s3:::{s3_bucket_name}/*",
                            ],
                            actions=["s3:*"],
                        ),
                        iam.PolicyStatement(
                            sid="DenyDeleteBucket",
                            effect=iam.Effect.DENY,
                            resources=[f"arn:aws:s3:::{s3_bucket_name}"],
                            actions=["s3:DeleteBucket*", "s3:CreateBucket"],
                        ),
                    ]
                ),
            },
        )

        # Batch Instance Profile
        self.batch_instance_profile = iam.InstanceProfile(
            self,
            "BatchInstanceProfile",
            role=self.batch_instance_role,
        )

        # Batch Spot Fleet Role
        self.batch_spot_fleet_role = iam.Role(
            self,
            "BatchSpotFleetRole",
            assumed_by=iam.ServicePrincipal("spotfleet.amazonaws.com"),
            managed_policies=[
                iam.ManagedPolicy.from_aws_managed_policy_name(
                    "service-role/AmazonEC2SpotFleetTaggingRole"
                ),
            ],
        )

        # Batch Service Role
        self.batch_service_role = iam.Role(
            self,
            "BatchServiceRole",
            assumed_by=iam.ServicePrincipal("batch.amazonaws.com"),
            managed_policies=[
                iam.ManagedPolicy.from_aws_managed_policy_name(
                    "service-role/AWSBatchServiceRole"
                ),
            ],
        )

        # Outputs
        cdk.CfnOutput(self, "BatchJobRoleArn", value=self.batch_job_role.role_arn)
        cdk.CfnOutput(
            self, "BatchServiceRoleArn", value=self.batch_service_role.role_arn
        )
        cdk.CfnOutput(
            self, "BatchSpotFleetRoleArn", value=self.batch_spot_fleet_role.role_arn
        )
        cdk.CfnOutput(
            self,
            "BatchInstanceProfileArn",
            value=self.batch_instance_profile.instance_profile_arn,
        )

File: playbook/ways-to-run/patterns/batch-squared/infrastructure/typescript/lib/iam-stack.ts
import * as cdk from "aws-cdk-lib";
import * as iam from "aws-cdk-lib/aws-iam";
import type { Construct } from "constructs";

export interface IamStackProps extends cdk.NestedStackProps {
  s3BucketName: string;
}

export class IamStack extends cdk.NestedStack {
  public readonly batchJobRole: iam.Role;
  public readonly batchInstanceRole: iam.Role;
  public readonly batchInstanceProfile: iam.InstanceProfile;
  public readonly batchSpotFleetRole: iam.Role;
  public readonly batchServiceRole: iam.Role;

  constructor(scope: Construct, id: string, props: IamStackProps) {
    super(scope, id, props);

    const region = cdk.Stack.of(this).region;

    // Batch Job Role
    this.batchJobRole = new iam.Role(this, "BatchJobRole", {
      assumedBy: new iam.ServicePrincipal("ecs-tasks.amazonaws.com"),
      inlinePolicies: {
        [`S3Bucket-Access-${region}`]: new iam.PolicyDocument({
          statements: [
            new iam.PolicyStatement({
              effect: iam.Effect.DENY,
              resources: [`arn:aws:s3:::${props.s3BucketName}`],
              actions: ["s3:Delete*", "s3:PutBucket*"],
            }),
            new iam.PolicyStatement({
              effect: iam.Effect.ALLOW,
              resources: [`arn:aws:s3:::${props.s3BucketName}`],
              actions: ["s3:ListBucket*"],
            }),
            new iam.PolicyStatement({
              effect: iam.Effect.ALLOW,
              resources: [`arn:aws:s3:::${props.s3BucketName}/*`],
              actions: ["s3:*"],
            }),
          ],
        }),
        // Read access to public data buckets used by nf-core pipelines
        "PublicDataBuckets-ReadOnly": new iam.PolicyDocument({
          statements: [
            new iam.PolicyStatement({
              sid: "AllowListPublicDataBuckets",
              effect: iam.Effect.ALLOW,
              resources: [
                "arn:aws:s3:::pgp10-fastqs",
                "arn:aws:s3:::ngi-igenomes",
                "arn:aws:s3:::1000genomes",
              ],
              actions: ["s3:ListBucket*"],
            }),
            new iam.PolicyStatement({
              sid: "AllowGetPublicDataObjects",
              effect: iam.Effect.ALLOW,
              resources: [
                "arn:aws:s3:::pgp10-fastqs/*",
                "arn:aws:s3:::ngi-igenomes/*",
                "arn:aws:s3:::1000genomes/*",
              ],
              actions: ["s3:GetObject", "s3:GetObjectVersion"],
            }),
          ],
        }),
      },
    });

    // Batch Instance Role
    this.batchInstanceRole = new iam.Role(this, "BatchInstanceRole", {
      assumedBy: new iam.ServicePrincipal("ec2.amazonaws.com"),
      managedPolicies: [
        iam.ManagedPolicy.fromAwsManagedPolicyName(
          "service-role/AmazonEC2ContainerServiceforEC2Role",
        ),
        iam.ManagedPolicy.fromAwsManagedPolicyName("AmazonS3ReadOnlyAccess"),
        iam.ManagedPolicy.fromAwsManagedPolicyName(
          "AmazonSSMManagedInstanceCore",
        ),
        iam.ManagedPolicy.fromAwsManagedPolicyName(
          "CloudWatchAgentServerPolicy",
        ),
      ],
      inlinePolicies: {
        [`S3Bucket-Access-${region}`]: new iam.PolicyDocument({
          statements: [
            new iam.PolicyStatement({
              sid: "S3BucketAllowAllObjectOps",
              effect: iam.Effect.ALLOW,
              resources: [
                `arn:aws:s3:::${props.s3BucketName}`,
                `arn:aws:s3:::${props.s3BucketName}/*`,
              ],
              actions: ["s3:*"],
            }),
            new iam.PolicyStatement({
              sid: "DenyDeleteBucket",
              effect: iam.Effect.DENY,
              resources: [`arn:aws:s3:::${props.s3BucketName}`],
              actions: ["s3:DeleteBucket*", "s3:CreateBucket"],
            }),
          ],
        }),
      },
    });

    // Batch Instance Profile
    this.batchInstanceProfile = new iam.InstanceProfile(
      this,
      "BatchInstanceProfile",
      {
        role: this.batchInstanceRole,
      },
    );

    // Batch Spot Fleet Role
    this.batchSpotFleetRole = new iam.Role(this, "BatchSpotFleetRole", {
      assumedBy: new iam.ServicePrincipal("spotfleet.amazonaws.com"),
      managedPolicies: [
        iam.ManagedPolicy.fromAwsManagedPolicyName(
          "service-role/AmazonEC2SpotFleetTaggingRole",
        ),
      ],
    });

    // Batch Service Role
    this.batchServiceRole = new iam.Role(this, "BatchServiceRole", {
      assumedBy: new iam.ServicePrincipal("batch.amazonaws.com"),
      managedPolicies: [
        iam.ManagedPolicy.fromAwsManagedPolicyName(
          "service-role/AWSBatchServiceRole",
        ),
      ],
    });

    // Outputs
    new cdk.CfnOutput(this, "BatchJobRoleArn", {
      value: this.batchJobRole.roleArn,
    });

    new cdk.CfnOutput(this, "BatchServiceRoleArn", {
      value: this.batchServiceRole.roleArn,
    });

    new cdk.CfnOutput(this, "BatchSpotFleetRoleArn", {
      value: this.batchSpotFleetRole.roleArn,
    });

    new cdk.CfnOutput(this, "BatchInstanceProfileArn", {
      value: this.batchInstanceProfile.instanceProfileArn,
    });
  }
}
